compiler/rustc_passes/src/check_export.rs

```diff
@@ -0,0 +1,397 @@
+/// 2. Primitive types.
+/// 3. Non-generic functions with a stable ABI (e.g. extern "C") for which every user
+///    defined type used in the signature is also marked as `#[export]`.
+fn exportable_items_provider_local<'tcx>(tcx: TyCtxt<'tcx>, _: LocalCrate) -> &'tcx [DefId] {
+    let mut visitor = ExportableItemCollector::new(tcx);
+    tcx.hir_walk_toplevel_module(&mut visitor);
+    let exportable_items = visitor.exportable_items;
+    for item_id in exportable_items.iter() {
+        let mut validator =
+            ExportableItemsChecker { tcx, exportable_items: &exportable_items, item_id: *item_id };
+        validator.check();
+    }
+
+    tcx.arena.alloc_from_iter(exportable_items.into_iter())
+}
+
+struct ImplsOrderVisitor<'tcx> {
+    tcx: TyCtxt<'tcx>,
+    order: FxIndexMap<DefId, usize>,
+    id: usize,
+}
+
+impl<'tcx> ImplsOrderVisitor<'tcx> {
+    fn new(tcx: TyCtxt<'tcx>) -> ImplsOrderVisitor<'tcx> {
+        ImplsOrderVisitor { tcx, order: Default::default(), id: 0 }
+    }
+}
+
+impl<'tcx> Visitor<'tcx> for ImplsOrderVisitor<'tcx> {
+    type NestedFilter = nested_filter::All;
+
+    fn maybe_tcx(&mut self) -> Self::MaybeTyCtxt {
+        self.tcx
+    }
+
+    fn visit_item(&mut self, item: &'tcx hir::Item<'tcx>) {
+        if let hir::ItemKind::Impl(impl_) = item.kind
+            && impl_.of_trait.is_none()
+            && self.tcx.is_exportable(item.owner_id.def_id.to_def_id())
+        {
+            self.order.insert(item.owner_id.def_id.to_def_id(), self.id);
+            self.id += 1;
+        }
+        intravisit::walk_item(self, item);
+    }
+}
+
+/// During symbol mangling rustc uses a special index to distinguish between two impls of
+/// the same type in the same module(See `DisambiguatedDefPathData`). For exportable items
+/// we cannot use the current approach because it is dependent on the compiler's
+/// implementation.
+///
+/// In order to make disambiguation independent of the compiler version we can assign an
+/// id to each impl according to the order of elements in the source code. However, there is
+/// no "proper" way to get this id, so we use:
+///   1. The sequential number of an impl during the `intravisit::Visitor` traversal.
+///   2. A new attribute `#[rustc_stable_impl_id]` that outputs the impl id as a compiler error.
+///      If the visitor's implementation is changed, the corresponding test will fail. Then
+///      you will need to rewrite the implementation of this query to preserve order.
+fn stable_order_of_exportable_impls<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    _: LocalCrate,
+) -> &'tcx FxIndexMap<DefId, usize> {
+    let mut vis = ImplsOrderVisitor::new(tcx);
+    tcx.hir_walk_toplevel_module(&mut vis);
+    tcx.arena.alloc(vis.order)
+}
+
+pub(crate) fn provide(providers: &mut Providers) {
+    *providers = Providers {
+        exportable_items: exportable_items_provider_local,
+        stable_order_of_exportable_impls,
+        ..*providers
+    };
+}
```

Comment:
I didn't look at the logic in this file in detail, it 100% has issues, but it should be ok approximation for a first prototype.